File: src/058/p058.go
package problem058

import (
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math/primeutils"
)

func Solve() string {
	primes := primeutils.GeneratePrimes(30000)
	last := 1
	countNumbers := 1 // 1 is already counted
	countPrimes := 0
	for side := 3; ; side += 2 {
		for i := 0; i < 3; i++ {
			last += side - 1
			if primeutils.IsPrime(last, primes) {
				countPrimes++
			}
		}
		countNumbers += 4
		if countPrimes*10 < countNumbers {
			return strconv.Itoa(side)
		}
		last += side - 1
	}
}

File: src/P058/p058.py
import itertools
from src.utils.py.prime import generate_primes, is_prime


def solve() -> int:
    primes = generate_primes(30000)
    count = 1
    count_primes = 0

    last = 1
    for sidelen in itertools.count(start=3, step=2):
        for _ in range(3):
            last += sidelen - 1
            if is_prime(last, primes):
                count_primes += 1
        count += 4
        last += sidelen - 1
        if count_primes / count < 0.1:
            return sidelen
    raise RuntimeError("No solution found")
